pageserver/src/tenant/timeline/walreceiver/connection_manager.rs

```diff
@@ -214,6 +227,61 @@ pub(super) async fn connection_manager_loop_step(
                     }
                 }
             } => debug!("Waking up for the next retry after waiting for {time_until_next_retry:?}"),
+
+            Some(()) = async {
+                // Reminder: this match arm needs to be cancellation-safe.
+                // Calculating time needed to wait until sending the next discovery request.
+                // Current implementation is conservative and sends discovery requests only when there are no candidates.
+
+                if any_activity {
+                    // No need to send discovery requests if there is an active connection or candidates.
+                    return None;
+                }
+
+                // Waiting for an active wait_lsn request.
+                while wait_lsn_status.borrow().is_none() {
+                    wait_lsn_status.changed().await.ok();
```

Comment:
This is a tight infinite loop without a yield to executor in it, when the sending side of the `broadcast` has gone away.

I suppose it can never happen because `Self` holds a strong reference to the Timeline, which in turn contains the sending side of the `wait_lsn_status`.
But, that's a subtlelty that can easily be missed in refactorings.

I suggest that if `changed()` returns an error, log a `warn!` about the sender having gone away, and invalidate the `wait_lsn_status` after that. Perhaps by putting the `wait_lsn_status` into an `Option` itself.